unofficial.gni

```diff
@@ -0,0 +1,287 @@
+    if (is_mac) {
+      frameworks = [ "CoreFoundation.framework" ]
+    }
+
+    if (v8_enable_i18n_support) {
+      deps += [ "//third_party/icu" ]
+    }
+    if (node_use_openssl) {
+      public_deps += [ "deps/openssl" ]
+      sources += gypi_values.node_crypto_sources
+    }
+    if (node_enable_inspector) {
+      deps += [
+        "src/inspector:node_protocol_generated_sources",
+        "src/inspector:v8_inspector_compress_protocol_json",
+      ]
+      include_dirs = [
+        "$target_gen_dir",
+        "$target_gen_dir/src",
+        "$target_gen_dir/src/inspector",
+      ]
+      node_inspector = exec_script(
+          "./tools/gypi_to_gn.py",
+          [ rebase_path("src/inspector/node_inspector.gypi"),
+            "--replace=<(SHARED_INTERMEDIATE_DIR)=$target_gen_dir" ],
+          "scope",
+          [ "src/inspector/node_inspector.gypi" ])
+      sources += node_inspector.node_inspector_sources +
+                 node_inspector.node_inspector_generated_sources
+    }
+  }
+
+  executable(target_name) {
+    forward_variables_from(invoker, "*")
+
+    sources = [ "src/node_main.cc" ]
+    deps = [ ":libnode" ]
+    if (node_use_node_snapshot) {
+      sources += [ "$target_gen_dir/node_snapshot.cc" ]
+      deps += [ ":run_node_mksnapshot" ]
+      if (is_clang || !is_win) {
+        cflags_cc = [
+          "-Wno-c++11-narrowing",
+          "-Wno-shadow",
+        ]
+      }
+    } else {
+      sources += [ "src/node_snapshot_stub.cc" ]
+    }
+    output_name = "node"
+  }
+
+  if (node_use_node_snapshot) {
+    if (current_toolchain == v8_snapshot_toolchain) {
+      executable("node_mksnapshot") {
+        configs += [ ":node_internal_config" ]
+        sources = [
+          "src/node_snapshot_stub.cc",
+          "tools/snapshot/node_mksnapshot.cc",
+        ]
+        deps = [ ":libnode" ]
+      }
+    }
+
+    action("run_node_mksnapshot") {
+      deps = [ ":node_mksnapshot($v8_snapshot_toolchain)" ]
+      script = "$node_v8_path/tools/run.py"
+      sources = []
+      data = []
+
+      mksnapshot_dir = get_label_info(":node_mksnapshot($v8_snapshot_toolchain)",
+                                      "root_out_dir")
+
+      outputs = [ "$target_gen_dir/node_snapshot.cc" ]
+      args = [
+        "./" + rebase_path(mksnapshot_dir + "/node_mksnapshot", root_build_dir),
+        rebase_path("$target_gen_dir/node_snapshot.cc", root_build_dir),
+      ]
+    }
+  }
+
+  action("generate_config_gypi") {
+    script = "./tools/generate_config_gypi.py"
+    outputs = [ "$target_gen_dir/config.gypi" ]
+    depfile = "$target_gen_dir/$target_name.d"
+    script_args = [ "$root_build_dir" ]
+    script_args += outputs
+    script_args += [ depfile ]
+    args = rebase_path(script_args, root_build_dir)
+  }
+
+  action("node_js2c") {
+    script = "tools/js2c.py"
+    deps = [ ":generate_config_gypi" ]
+
+    # TODO(zcbenz): These files are defined in configure.py.
+    node_deps_files = gypi_values.deps_files + [
```

Comment:
`configure.py` would write these files to the `node_builtin_shareable_builtins` variable in `./config.gypi` after it's run. I wonder if we could hook into the workflow of `configure.py` so that instead of running GYP, it runs  GN with an option like `--build-with-gn` (which I think was what V8 used to do at some point). Then the GN files, consumed when `configure.py` runs GN, can get those variables from `./config.gypi`. That might also make setting up the CI a bit easier if the workflow isn't very different from the GYP one:

```
./configure --build-with-gn
BUILD_WITH=ninja make
```

And it would be a bit more robust than the `generate_config_gypi.py` approach.

Reply:
GN works very similar to GYP (i.e. `gn gen --args out/Release && ninja -C out/Release`), however it does not ship any preset build configurations with it, thus GN does not know how to invoke C++ compilers by default, and to build Node with GN we have to checkout Chromium's build configuration files, which are quite heavy (you can check https://github.com/photoionization/node-ci/blob/main/DEPS for example).

So in order to support running GN in `configure.py`, we have to ship a bunch of Chromium's tools inside Node too, which is a huge maintenance burden, and we need to update those files whenever V8 is updated.

Reply:
I am actually fine with adding some scripts like https://github.com/photoionization/node_with_gn/tree/gn-in-node/scripts to this repo to make it work - not a full checkout in tree, but some revs to download from is probably good enough (if the purpose is for others to always have a working GN build in this repo, I suppose it would be good to keep it in sync with the tool updates). But I understand that it could be better to do that in a follow-up to avoid complicating this one.